File: python3/determine_if_string_halves_are_alike.py
"""
Space   : O(1)
Time	: O(n)
"""


class Solution:
    def halvesAreAlike(self, s: str) -> bool:
        vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
        n = len(s)

        start = n // 2
        end = n // 2
        if n % 2 == 0:
            start -= 1

        h1, h2 = 0, 0
        while start >= 0:
            if s[start] in vowels:
                h1 += 1
            
            if s[end] in vowels:
                h2 += 1

            start -= 1
            end += 1
        
        return h1 == h2
        

File: golang/determine_if_string_halves_are_alike.go
package main

// Space 	: O(1)
// Time	    : O(n)

import (
	"unicode"
)

func halvesAreAlike(s string) bool {
	v1, v2 := 0, 0
	n := len(s)
	vowels := map[rune]int{
		'a': 1,
		'e': 1,
		'i': 1,
		'o': 1,
		'u': 1,
	}

	for i := 0; i < n/2; i++ {
		if _, ok := vowels[unicode.ToLower(rune(s[i]))]; ok {
			v1++
		}
		if _, ok := vowels[unicode.ToLower(rune(s[i+(n/2)]))]; ok {
			v2++
		}

	}

	return v1 == v2
}
